data_pipeline/generic_web_api/generic_web_api_data_etl.py

```diff
@@ -0,0 +1,448 @@
+    return item_list
+
+
+def upload_latest_timestamp(data_config, latest_record_timestamp: datetime):
+    if (
+            data_config.state_file_object_name and
+            data_config.state_file_bucket_name
+    ):
+        latest_record_date = get_timestamp_as_string(latest_record_timestamp)
+        state_file_name_key = data_config.state_file_object_name
+        state_file_bucket = data_config.state_file_bucket_name
+        upload_s3_object(
+            bucket=state_file_bucket,
+            object_key=state_file_name_key,
+            data_object=latest_record_date,
+        )
+
+
+def create_n_extend_table(
+        data_config: WebApiConfig,
+        full_temp_file_location,
+):
+    schema = generate_schema_from_file(
+        full_temp_file_location
+    )
+
+    if does_bigquery_table_exist(
+            data_config.gcp_project,
+            data_config.dataset_name,
+            data_config.table_name,
+    ):
+        extend_table_schema_with_nested_schema(
+            data_config.gcp_project,
+            data_config.dataset_name,
+            data_config.table_name,
+            schema
+        )
+    else:
+        create_table(
+            data_config.gcp_project,
+            data_config.dataset_name,
+            data_config.table_name,
+            schema
+        )
+
+
+def get_bq_schema(data_config: WebApiConfig,):
+    if (
+            data_config.schema_file_object_name and
+            data_config.schema_file_s3_bucket
+    ):
+        bq_schema = download_s3_json_object(
+            data_config.schema_file_s3_bucket,
+            data_config.schema_file_object_name
+        )
+    else:
+        bq_schema = None
+    return bq_schema
+
+
+def process_downloaded_data(
+        record_list: list,
+        data_config: WebApiConfig,
+        data_etl_timestamp,
+        file_location,
+        prev_page_latest_timestamp: datetime = None
+):
+    provenance = {
+        data_config.import_timestamp_field_name:
+            data_etl_timestamp
+    }
+    bq_schema = get_bq_schema(data_config)
+    n_record_list = process_record_in_list(
+        record_list=record_list, bq_schema=bq_schema,
+        provenance=provenance
+    )
+    n_record_list = write_result_to_file(
+        n_record_list, file_location
+    )
+    current_page_latest_timestamp = get_latest_record_list_timestamp(
+        n_record_list, prev_page_latest_timestamp, data_config
+    )
+    return current_page_latest_timestamp
+
+
+def process_record_in_list(
+        record_list,
+        provenance: dict = None,
+        bq_schema=None
+) -> Iterable:
+    for record in record_list:
+        n_record = standardize_record_keys(record)
+        if bq_schema:
+            n_record = filter_record_by_schema(n_record, bq_schema)
+        if provenance:
+            n_record.update(provenance)
+        yield n_record
+
+
+# pylint: disable=inconsistent-return-statements,broad-except,no-else-return
+def standardize_record_keys(record_object):
```

Comment:
question: did we not already have similar BigQuery and normalization related methods in other pipelines?

Reply:
in a way, no, csv headers are standardized before use. `crossref event data pipeline` has similar functionality but this functionality is mixed mixed with  response record filtering based on the bigquery table schema.